server/src/test/java/org/elasticsearch/cluster/service/MasterServiceTests.java

```diff
@@ -982,7 +992,12 @@ public void onResponse(ClusterState clusterState) {
 
                         @Override
                         public void onFailure(Exception e) {
-                            throw new AssertionError(e);
+                            assertTrue(expectFailure);
+                            assertThat(e, instanceOf(ElasticsearchException.class));
+                            assertThat(e.getMessage(), equalTo(exceptionMessage));
+                            assertEquals(List.of(testResponseHeaderValue), threadContext.getResponseHeaders().get(testResponseHeaderName));
+                            assertNotNull(publishedState.get());
+                            publishSuccessCountdown.countDown();
```

Comment:
Can we assert that `publishSuccessCountdown.getCount() > 0`? Also in `onSuccess`.